### ISSUE TITLE
The timer runs one more time unnecessarily in the background whenever closing and reopening the app

### ISSUE THREAD

**Issue by Yonghwan-Song** [Yonghwan-Song/pomodoro#6]
# Problem
## When timer is running  in the background  **by counting down the duration every 0.5sec**
![Screenshot from 2023-06-11 12-16-44](https://github.com/Yonghwan-Song/pomodoro/assets/72689705/d2aa1712-9122-499a-b91a-0f619e8338a8)

## After log-out and log-in
![image](https://github.com/Yonghwan-Song/pomodoro/assets/72689705/277b9c88-9b45-4b41-ad65-466ac3977f95)

Timer is still counting down after log-out. After signing in again, page is redirected to "/main" and the interval in the background is cleared like below.

![Screenshot from 2023-06-11 12-31-36](https://github.com/Yonghwan-Song/pomodoro/assets/72689705/7df2a4c3-b967-447a-8312-53da7c0d0938)

## After closing and re-opeing the app while timer is running in the background
The timer is run by two `setInterval` functions every 0.5 second.
In other words, the `setInterval` function is called one more time unnecessarily.
![image](https://github.com/Yonghwan-Song/pomodoro/assets/72689705/9729e04e-e8c4-481e-81cd-5edd37b123f4)

**Comment by Yonghwan-Song** [comment]
# Cause

1. Signing out did not affect the lifecycle of the App component since it was for the root route "/"
2. But when closing and re-opening the app, App was mounted and "sendDataToIndex" message was sent.
   **App.tsx**

```typescript
...

 useEffect(() => {
    console.log("APP is being mounted");

    SW?.postMessage("sendDataToIndex"); // <----------------

    function onUnload() {
      localStorage.removeItem("isTimerRunning");
    }
    window.addEventListener("unload", onUnload);
     
    return () => {
      window.removeEventListener("unload", onUnload);
    };
  }, []);

...
```

3. In the service worker, `countDown` function was called unnecessarily because the if conditional statement was not thorough. **There was no conditional statement to check whether or not the timer is already running in the background.**

```javascript
...

} else if (ev.data === "sendDataToIndex") {
    //! This is when a user is leaving a main page(origin/main).
    //! Thus, from this point, this service worker is responsible for counting down the timer.
    if (DB) {
      sendStates(ev.source.id).then((states) => {
        if (states.running) { <------------------------------
          countDown(states, ev.source.id);
        }
      });
    } else {
      openDB(() => {
        sendStates(ev.source.id).then((states) => {
          if (states.running) { <----------------------------
            countDown(states, ev.source.id);
          }
        });
      });
    }
  }

...

```

# Solution

1. Edit the `postMessage` call in the App.tsx

   ```typescript
   ...
   
       SW?.postMessage({
         action: "sendDataToIndex",
         payload: localStorage.getItem("idOfSetInterval"),
       });
   
   ...
   
   ```

   

2. Add a new condition in the if conditional statement in the service worker

   This file is not yet bundled meaning it is different than the `sw2.js`

   ```javascript
   ...
   
    if (ev.data.action && ev.data.action === "sendDataToIndex") {
      if (DB) {
        sendStates(ev.source.id).then((states) => {
          if (states.running && ev.data.payload === null) {
            countDown(states, ev.source.id);
          }
        });
      } else {
        openDB(() => {
          sendStates(ev.source.id).then((states) => {
            if (states.running && ev.data.payload === null) {
              countDown(states, ev.source.id);
            }
          });
        });
      }
    }
   
   ...
   
   ```
